File: src/astrodyn_core/geqoe_cpp/src/conversions.cpp
#include "conversions.hpp"
#include "kepler_solver.hpp"

#include <cmath>
#include <stdexcept>
#include <vector>

namespace astrodyn_core {
namespace geqoe {

// ============================================================================
// rv2geqoe  --  Cartesian -> GEqOE (per Giulio Bau's formulation with J2)
// ============================================================================
void rv2geqoe(
    const double* y_in,
    double* eq_out,
    size_t N,
    double J2, double Re, double mu
) {
    // Normalisation constants
    const double L     = Re;
    const double T     = std::sqrt(Re * Re * Re / mu);
    const double inv_L = 1.0 / L;
    const double T_over_L = T / L;           // = T * inv_L
    // mu_norm = 1.0 (implicit -- all mu_norm factors are unity)
    const double A     = J2 / 2.0;
    const double inv_T = 1.0 / T;

    for (size_t i = 0; i < N; ++i) {
        const double* y = y_in + i * 6;
        double* eq      = eq_out + i * 6;

        // --- Normalised position and velocity ---
        double rx = y[0] * inv_L;
        double ry = y[1] * inv_L;
        double rz = y[2] * inv_L;
        double vx = y[3] * T_over_L;
        double vy = y[4] * T_over_L;
        double vz = y[5] * T_over_L;

        // --- Geometric quantities ---
        double r2 = rx * rx + ry * ry + rz * rz;
        double r  = std::sqrt(r2);
        double inv_r = 1.0 / r;
        double rp = (rx * vx + ry * vy + rz * vz) * inv_r;   // radial velocity
        double v2 = vx * vx + vy * vy + vz * vz;

        // --- J2 potential ---
        double zg  = rz * inv_r;
        double r3  = r * r2;
        double U   = -A / r3 * (1.0 - 3.0 * zg * zg);

        // --- Radial unit vector ---
        double erx = rx * inv_r, ery = ry * inv_r, erz = rz * inv_r;

        // --- Angular momentum  h = r x v ---
        double hx = ry * vz - rz * vy;
        double hy = rz * vx - rx * vz;
        double hz = rx * vy - ry * vx;
        double h2 = hx * hx + hy * hy + hz * hz;
        double h  = std::sqrt(h2);
        double inv_h = 1.0 / h;

        double Ueff = h2 / (2.0 * r2) + U;

        // --- Angular momentum unit vector ---
        double ehx = hx * inv_h, ehy = hy * inv_h, ehz = hz * inv_h;

        // --- In-plane perpendicular unit vector  ef = eh x er ---
        double efx = ehy * erz - ehz * ery;
        double efy = ehz * erx - ehx * erz;
        double efz = ehx * ery - ehy * erx;

        // --- Total energy & mean motion ---
        double E_val   = v2 / 2.0 - 1.0 / r + U;    // mu_norm = 1
        double neg2E   = -2.0 * E_val;
        double nu_norm = std::pow(neg2E, 1.5);        // (1/mu) * (-2E)^1.5, mu=1

        // --- Orientation parameters q1, q2 ---
        // eh . ez = ehz,  eh . ex = ehx,  eh . ey = ehy
        double ehez = ehz;
        double inv_1pehez = 1.0 / (1.0 + ehez);
        double q1 =  ehx * inv_1pehez;
        double q2 = -ehy * inv_1pehez;

        // --- Equinoctial frame vectors eX, eY ---
        double q1s = q1 * q1, q2s = q2 * q2, q1q2 = q1 * q2;
        double eTerm = 1.0 / (1.0 + q1s + q2s);

        double eXx = eTerm * (1.0 - q1s + q2s);
        double eXy = eTerm * (2.0 * q1q2);
        double eXz = eTerm * (-2.0 * q1);

        double eYx = eTerm * (2.0 * q1q2);
        double eYy = eTerm * (1.0 + q1s - q2s);
        double eYz = eTerm * (2.0 * q2);

        // --- Generalised eccentricity vector ---
        double c = std::sqrt(2.0 * r2 * Ueff);
        double c_over_r = c * inv_r;

        // gv = rp * er + (c/r) * ef
        double gvx = rp * erx + c_over_r * efx;
        double gvy = rp * ery + c_over_r * efy;
        double gvz = rp * erz + c_over_r * efz;

        // cross1 = rv x gv
        double c1x = ry * gvz - rz * gvy;
        double c1y = rz * gvx - rx * gvz;
        double c1z = rx * gvy - ry * gvx;

        // g = gv x cross1 - er   (mu_norm = 1)
        double gx = (gvy * c1z - gvz * c1y) - erx;
        double gy = (gvz * c1x - gvx * c1z) - ery;
        double gz = (gvx * c1y - gvy * c1x) - erz;

        // --- Eccentricity-like parameters ---
        double p1_val = gx * eYx + gy * eYy + gz * eYz;    // g . eY
        double p2_val = gx * eXx + gy * eXy + gz * eXz;    // g . eX

        // --- Position projections onto equinoctial frame ---
        double X = rx * eXx + ry * eXy + rz * eXz;         // rv . eX
        double Y = rx * eYx + ry * eYy + rz * eYz;         // rv . eY

        // --- True longitude ---
        double p1s_val = p1_val * p1_val;
        double p2s_val = p2_val * p2_val;
        double beta     = std::sqrt(1.0 - p1s_val - p2s_val);
        double alpha    = 1.0 / (1.0 + beta);
        // Match Python: (mu_norm / nu_norm**2) ** (1/3)  with mu_norm=1
        double a        = std::pow(1.0 / (nu_norm * nu_norm), 1.0 / 3.0);
        double inv_a_beta = 1.0 / (a * beta);

        double cosK_val = p2_val + inv_a_beta * ((1.0 - alpha * p2s_val) * X
                                                - alpha * p1_val * p2_val * Y);
        double sinK_val = p1_val + inv_a_beta * ((1.0 - alpha * p1s_val) * Y
                                                - alpha * p1_val * p2_val * X);

        double Lr = std::atan2(sinK_val, cosK_val)
                  + inv_a_beta * (X * p1_val - Y * p2_val);

        // --- Output (denormalise nu) ---
        eq[0] = nu_norm * inv_T;
        eq[1] = q1;
        eq[2] = q2;
        eq[3] = p1_val;
        eq[4] = p2_val;
        eq[5] = Lr;
    }
}

// ============================================================================
// geqoe2rv  --  GEqOE -> Cartesian
// ============================================================================
void geqoe2rv(
    const double* eq_in,
    double* rv_out,
    double* rpv_out,
    size_t N,
    double J2, double Re, double mu
) {
    const double L        = Re;
    const double T        = std::sqrt(Re * Re * Re / mu);
    const double inv_T    = 1.0 / T;
    const double L_over_T = L * inv_T;
    const double A        = J2 / 2.0;
    const double TWO_PI   = 2.0 * M_PI;
    const double ONE_THIRD = 1.0 / 3.0;

    // --- Vectorised Kepler solve (matches Python np.all convergence) --------
    // Extract Lr, p1, p2 into temporary arrays, call the vectorised solver
    // so the convergence path is identical to the Python reference.
    std::vector<double> Lr_vec(N), p1_vec(N), p2_vec(N), K_vec(N);
    std::vector<double> nu_vec(N);

    for (size_t i = 0; i < N; ++i) {
        const double* eq = eq_in + i * 6;
        nu_vec[i] = eq[0] * T;

        double Lr_mod = std::fmod(eq[5], TWO_PI);
        if (Lr_mod < 0.0) Lr_mod += TWO_PI;
        Lr_vec[i] = Lr_mod;

        p1_vec[i] = eq[3];
        p2_vec[i] = eq[4];
    }

    solve_kep_gen(Lr_vec.data(), p1_vec.data(), p2_vec.data(),
                  K_vec.data(), N);

    // --- Per-state Cartesian reconstruction ---------------------------------
    for (size_t i = 0; i < N; ++i) {
        const double* eq = eq_in + i * 6;
        double* rv  = rv_out  + i * 3;
        double* rpv = rpv_out + i * 3;

        double nu_norm = nu_vec[i];
        double q1 = eq[1], q2 = eq[2];
        double p1 = p1_vec[i], p2 = p2_vec[i];

        double K    = K_vec[i];
        double sinK = std::sin(K);
        double cosK = std::cos(K);

        // Equinoctial frame
        double q1s = q1 * q1, q2s = q2 * q2, q1q2 = q1 * q2;
        double eTerm = 1.0 / (1.0 + q1s + q2s);

        double eXx = eTerm * (1.0 - q1s + q2s);
        double eXy = eTerm * (2.0 * q1q2);
        double eXz = eTerm * (-2.0 * q1);

        double eYx = eTerm * (2.0 * q1q2);
        double eYy = eTerm * (1.0 + q1s - q2s);
        double eYz = eTerm * (2.0 * q2);

        // Orbital shape
        double p1s_val = p1 * p1, p2s_val = p2 * p2;
        double gs    = p1s_val + p2s_val;
        double beta  = std::sqrt(1.0 - gs);
        double alpha = 1.0 / (1.0 + beta);
        // Match Python: (mu_norm / nu_norm**2) ** (1/3)  with mu_norm=1
        double a     = std::pow(1.0 / (nu_norm * nu_norm), ONE_THIRD);

        double X = a * (alpha * p1 * p2 * sinK + (1.0 - alpha * p1s_val) * cosK - p2);
        double Y = a * (alpha * p1 * p2 * cosK + (1.0 - alpha * p2s_val) * sinK - p1);

        // Normalised position
        double rvx = X * eXx + Y * eYx;
        double rvy = X * eXy + Y * eYy;
        double rvz = X * eXz + Y * eYz;

        double z = rvz;
        double r = a * (1.0 - p1 * sinK - p2 * cosK);
        double inv_r = 1.0 / r;
        // sqrt(mu_norm * a) / r * (...), mu_norm=1
        double rp_val = std::sqrt(a) * inv_r * (p2 * sinK - p1 * cosK);

        double cosL = X * inv_r;
        double sinL = Y * inv_r;
        // (mu_norm**2 / nu_norm) ** (1/3) * beta, mu_norm=1
        double c    = std::pow(1.0 / nu_norm, ONE_THIRD) * beta;

        // J2 potential at this position
        double zg  = z * inv_r;
        double r3  = r * r * r;
        double U   = -A / r3 * (1.0 - 3.0 * zg * zg);
        double h   = std::sqrt(c * c - 2.0 * r * r * U);
        double h_over_r = h * inv_r;

        // Normalised velocity
        double Xp = rp_val * cosL - h_over_r * sinL;
        double Yp = rp_val * sinL + h_over_r * cosL;

        double rpvx = Xp * eXx + Yp * eYx;
        double rpvy = Xp * eXy + Yp * eYy;
        double rpvz = Xp * eXz + Yp * eYz;

        // Denormalise to SI
        rv[0] = rvx * L;
        rv[1] = rvy * L;
        rv[2] = rvz * L;

        rpv[0] = rpvx * L_over_T;
        rpv[1] = rpvy * L_over_T;
        rpv[2] = rpvz * L_over_T;
    }
}

} // namespace geqoe
} // namespace astrodyn_core

File: src/astrodyn_core/geqoe_taylor/conversions.py
"""Cartesian <-> GEqOE conversions using K (eccentric longitude).

State vector: [nu, p1, p2, K, q1, q2] in physical units (km, s).
Reference: Baù et al. (2021), Sections 3-4, Appendix B.
"""

from __future__ import annotations

import numpy as np

from astrodyn_core.geqoe_taylor.perturbations.base import PerturbationModel


def cart2geqoe(
    r_vec: np.ndarray,
    v_vec: np.ndarray,
    mu: float,
    perturbation: PerturbationModel,
    t: float = 0.0,
) -> np.ndarray:
    """Convert Cartesian state to GEqOE with K as fast variable.

    Args:
        r_vec: position (3,) in km.
        v_vec: velocity (3,) in km/s.
        mu: gravitational parameter in km^3/s^2.
        perturbation: PerturbationModel providing U_numeric.
        t: epoch time in seconds (for time-dependent potentials).

    Returns:
        GEqOE state [nu, p1, p2, K, q1, q2].
    """
    r_vec = np.asarray(r_vec, dtype=float)
    v_vec = np.asarray(v_vec, dtype=float)

    # Geometric quantities
    r = np.linalg.norm(r_vec)
    v2 = np.dot(v_vec, v_vec)
    rdot = np.dot(r_vec, v_vec) / r

    # Angular momentum vector
    h_vec = np.cross(r_vec, v_vec)
    h = np.linalg.norm(h_vec)

    # Unit vectors
    er = r_vec / r
    eh = h_vec / h

    # Disturbing potential
    U = perturbation.U_numeric(r_vec, t)

    # Total energy (Eq. 16)
    E_val = v2 / 2.0 - mu / r + U

    # Generalized mean motion (Eq. 16 with a = (mu/nu^2)^(1/3))
    nu = (1.0 / mu) * (-2.0 * E_val) ** 1.5

    # Orientation parameters q1, q2 (Eq. 35-36)
    ex = np.array([1.0, 0.0, 0.0])
    ey = np.array([0.0, 1.0, 0.0])
    ez = np.array([0.0, 0.0, 1.0])

    eh_dot_ez = np.dot(eh, ez)
    denom = 1.0 + eh_dot_ez
    q1 = np.dot(eh, ex) / denom
    q2 = -np.dot(eh, ey) / denom

    # Equinoctial frame (Eq. 37)
    q1s = q1**2
    q2s = q2**2
    q1q2 = q1 * q2
    gamma_inv = 1.0 / (1.0 + q1s + q2s)

    eX = gamma_inv * np.array([1.0 - q1s + q2s, 2.0 * q1q2, -2.0 * q1])
    eY = gamma_inv * np.array([2.0 * q1q2, 1.0 + q1s - q2s, 2.0 * q2])

    # Effective angular momentum quantities (Eq. 6, 7, 23)
    Ueff = h**2 / (2.0 * r**2) + U
    c = r * np.sqrt(2.0 * Ueff)

    # Generalized radial velocity and force vector (Eq. 10)
    ef = np.cross(eh, er)
    upsilon_vec = rdot * er + (c / r) * ef

    # Generalized Laplace vector g (Eq. 10)
    g_vec = np.cross(upsilon_vec, np.cross(r_vec, upsilon_vec)) / mu - er

    # Eccentricity-like projections
    p1 = np.dot(g_vec, eY)
    p2 = np.dot(g_vec, eX)

    # Position projections onto equinoctial frame
    X = np.dot(r_vec, eX)
    Y = np.dot(r_vec, eY)

    # Compute sinK, cosK (Eq. 39 inverted from Eq. 42)
    g2 = p1**2 + p2**2
    beta = np.sqrt(1.0 - g2)
    alpha = 1.0 / (1.0 + beta)
    a = (mu / nu**2) ** (1.0 / 3.0)
    ab = a * beta

    cosK = p2 + (1.0 / ab) * ((1.0 - alpha * p2**2) * X - alpha * p1 * p2 * Y)
    sinK = p1 + (1.0 / ab) * ((1.0 - alpha * p1**2) * Y - alpha * p1 * p2 * X)

    K = np.arctan2(sinK, cosK)

    return np.array([nu, p1, p2, K, q1, q2])


def geqoe2cart(
    state: np.ndarray,
    mu: float,
    perturbation: PerturbationModel,
    t: float = 0.0,
) -> tuple[np.ndarray, np.ndarray]:
    """Convert GEqOE state to Cartesian.

    Args:
        state: GEqOE [nu, p1, p2, K, q1, q2].
        mu: gravitational parameter in km^3/s^2.
        perturbation: PerturbationModel providing U_numeric.
        t: epoch time in seconds.

    Returns:
        (r_vec, v_vec): position (3,) in km, velocity (3,) in km/s.
    """
    nu, p1, p2, K, q1, q2 = state

    # Equinoctial frame (Eq. 37)
    q1s = q1**2
    q2s = q2**2
    q1q2 = q1 * q2
    gamma_inv = 1.0 / (1.0 + q1s + q2s)

    eX = gamma_inv * np.array([1.0 - q1s + q2s, 2.0 * q1q2, -2.0 * q1])
    eY = gamma_inv * np.array([2.0 * q1q2, 1.0 + q1s - q2s, 2.0 * q2])

    # Shape quantities (Eq. 21, 40)
    g2 = p1**2 + p2**2
    beta = np.sqrt(1.0 - g2)
    alpha = 1.0 / (1.0 + beta)
    a = (mu / nu**2) ** (1.0 / 3.0)

    # Position from K (Eq. 42)
    sinK = np.sin(K)
    cosK = np.cos(K)

    X = a * (alpha * p1 * p2 * sinK + (1.0 - alpha * p1**2) * cosK - p2)
    Y = a * (alpha * p1 * p2 * cosK + (1.0 - alpha * p2**2) * sinK - p1)

    r_vec = X * eX + Y * eY

    # Distance and radial velocity (Eq. 31-32)
    r = a * (1.0 - p1 * sinK - p2 * cosK)
    rdot = np.sqrt(mu * a) / r * (p2 * sinK - p1 * cosK)

    # True longitude trig
    cosL = X / r
    sinL = Y / r

    # Generalized angular momentum (Eq. 23)
    c = (mu**2 / nu) ** (1.0 / 3.0) * beta

    # Physical angular momentum h from c and U (Eq. 44)
    U = perturbation.U_numeric(r_vec, t)
    h = np.sqrt(c**2 - 2.0 * r**2 * U)

    # Velocity components (Eq. 43)
    Xdot = rdot * cosL - (h / r) * sinL
    Ydot = rdot * sinL + (h / r) * cosL

    v_vec = Xdot * eX + Ydot * eY

    return r_vec, v_vec
